Cesium.CodeGen/Contexts/TranslationUnitContext.cs

```diff
@@ -115,7 +119,26 @@ internal void AddTypeDefinition(string name, IType type)
         _types.Add(name, type);
     }
 
-    internal void AddTagDefinition(string name, IType type) => _tags.Add(name, type);
+    internal void AddTagDefinition(string name, IType type)
+    {
+        if (_tags.TryGetValue(name, out var existingType))
+        {
+            if (type == existingType) return;
+            if (existingType.GetType() != type.GetType())
+            {
+                throw new CompilationException($"Tag kind {GetTypeKind(type.GetType())} type {name} was already defined as {GetTypeKind(existingType.GetType())}");
+            }
+        }
+
+        _tags.Add(name, type);
+
+        string GetTypeKind(Type type) => type.Name switch
+        {
+            nameof(StructType) => "struct",
+            nameof(EnumType) => "enum",
+            _ => throw new InvalidOperationException($"Unsupported type {type} used."),
+        };
```

Comment:
This fragment deserves some more attention.

1. Let's get rid of comparing `GetType()` directly.
2. Also, let's get rid of block like `GetTypeKind(Type type)` (I don't like that the `Type` gets passed around).

Instead, I'd prefer an `enum TypeKind` and a corresponding property in the `IType` interface.

Reply:
Based on what I see during implementation, and how we really construct type information.
It seems to me that instead of bunch of clases we can have 1 (probably) structure which repesents whole set of types.
```csharp
struct TypeInformation
{
    string name;
    TYpeKind Kind;
    bool IsPointer;
    TypeInformation * baseType;
    bool IsConstant;
    int Size;
    // Maybe some other properties.
}
```

That allow get rid of some type gymnastics which I have to do in some places already

Reply:
This would probably also be okay, though I'd prefer to review that as a separate PR.